### PR TITLE
feat: assign deployment error codes to deployment workflow

### PR DESCRIPTION
**Pull request by junfuchen99**
**Description of changes:**
1. Assign additional error codes to deployment workflow.
2. Save deployment error codes across nucleus restart during deployment.
3. Modified `DeploymentErrorCodeUtils::classifyComponentError` to use arn parser from AWS SDK.
4. Updated relevant tests.

**Why is this change necessary:**
To help service gain better visibility of deployment failures on device.
These changes are part of our ongoing improvements to FSS and device health notifications.

**How was this change tested:**
- [x] Updated or added new unit tests.
- [ ] Updated or added new integration tests.
- [ ] Updated or added new end-to-end tests.
- [ ] If my code makes a remote network call, it was tested with a proxy.

**Any additional information or context required to review the change:**
Integration tests are not updated because I have not made the changes to bubble up deployment error stack in deployment status details.
Will update them in the next pr.

**Documentation Checklist:**
 - [ ] Updated the README if applicable.

**Compatibility Checklist:**
- [x] I confirm that the change is backwards compatible.
- [ ] Any modification or deletion of public interfaces does not impact other plugin components.
- [ ] For external library version updates, I have reviewed its change logs and Nucleus does not consume 
  any deprecated method or type.

Refer to [Compatibility Guidelines](/COMPATIBILITY.md) for more information.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by MikeDombo** [comment]
We identified new issues on unchanged lines of code. Navigate to the Amazon CodeGuru Reviewer console to view the recommendations to fix them.

File: src/main/java/com/aws/greengrass/deployment/bootstrap/BootstrapManager.java [hunk 1]
```diff
@@ -400,7 +400,9 @@ protected int executeOneBootstrapTask(BootstrapTaskStatus next) throws ServiceUp
             next.setStatus(DONE);
             next.setExitCode(exitCode);
             return exitCode;
-        } catch (InterruptedException | TimeoutException | ServiceLoadException e) {
+        } catch (TimeoutException e) {
+            throw new ServiceUpdateException(e, DeploymentErrorCode.COMPONENT_BOOTSTRAP_TIMEOUT);
+        } catch (InterruptedException | ServiceLoadException e) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/bootstrap/BootstrapManager.java:405]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)

File: src/main/java/com/aws/greengrass/deployment/activator/DefaultActivator.java [hunk 2]
```diff
@@ -147,15 +146,15 @@ void rollback(DeploymentDocument deploymentDocument, CompletableFuture<Deploymen
...
+            waitForServicesToStart(servicesToTrackForRollback, mergeTime, componentStore);
 
             rollbackManager.removeObsoleteServices();
             logger.atInfo(MERGE_CONFIG_EVENT_KEY).kv(DEPLOYMENT_ID_LOG_KEY, deploymentId)
                     .log("All services rolled back");
 
             totallyCompleteFuture.complete(
                     new DeploymentResult(DeploymentResult.DeploymentStatus.FAILED_ROLLBACK_COMPLETE, failureCause));
-        } catch (InterruptedException | ExecutionException | ServiceUpdateException | ServiceLoadException e) {
+        } catch (InterruptedException | ServiceUpdateException | ServiceLoadException e) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/activator/DefaultActivator.java:157]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)

File: src/main/java/com/aws/greengrass/componentmanager/ComponentManager.java [hunk 3]
```diff
@@ -199,25 +207,22 @@ private ComponentIdentifier negotiateVersionWithCloud(String componentName,
                         "resolve-component-version", logger);
             } catch (InterruptedException e) {
                 throw e;
-            } catch (Exception e) {
+            } catch (NoAvailableComponentVersionException e) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/ComponentManager.java:210]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)

File: src/main/java/com/aws/greengrass/componentmanager/builtins/ArtifactDownloader.java [hunk 4]
```diff
@@ -131,12 +135,13 @@ public File download() throws PackageDownloadException, IOException, Interrupted
                     offset.set(0);
                     messageDigest.reset();
                     throw new ArtifactChecksumMismatchException(
-                            "Integrity check for downloaded artifact failed. " + "Probably due to file corruption.");
+                            "Integrity check for downloaded artifact failed. Probably due to file corruption.",
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/builtins/ArtifactDownloader.java:138]: let's fix this error message, it has been confusing people.

`Probably due to the file changing after creating the component version`

File: src/main/java/com/aws/greengrass/componentmanager/builtins/GreengrassRepositoryDownloader.java [hunk 5]
```diff
@@ -206,8 +218,19 @@ private String getArtifactDownloadURL(ComponentIdentifier componentIdentifier, S
             }, "get-artifact-size", logger);
         } catch (InterruptedException e) {
             throw e;
+        } catch (GreengrassV2DataException e) {
+            if (e.statusCode() == HttpStatusCode.FORBIDDEN) {
+                throw new PackageDownloadException(getErrorString("Failed to get the pre-signed url of a public or a "
+                        + "Lambda component artifact. GetComponentVersionArtifact returns 403 Access Denied"),
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/builtins/GreengrassRepositoryDownloader.java:224]: list the permission which is needed to execute this correctly, and note that it is in the iot policy.

File: src/main/java/com/aws/greengrass/componentmanager/builtins/S3Downloader.java [hunk 6]
```diff
@@ -111,15 +127,29 @@ public Long getDownloadSize() throws InterruptedException, PackageDownloadExcept
         // Parse artifact path
         String key = s3ObjectPath.key;
         String bucket = s3ObjectPath.bucket;
+        S3Client regionClient = getRegionClientForBucket(bucket);
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/builtins/S3Downloader.java:130]: this can fail, can't it? Why is this removed from the `try`?

File: src/main/java/com/aws/greengrass/componentmanager/builtins/S3Downloader.java [hunk 7]
```diff
@@ -142,6 +172,18 @@ private S3Client getRegionClientForBucket(String bucket) throws InterruptedExcep
                         message.substring(message.indexOf(REGION_EXPECTING_STRING) + REGION_EXPECTING_STRING.length());
                 region = message.substring(0, message.indexOf('\''));
             } else {
+                if (e.statusCode() == HttpStatusCode.FORBIDDEN) {
+                    throw new PackageDownloadException(getErrorString("Failed to determine S3 bucket location. "
+                            + "GetBucketLocation returns 403 Access Denied"), e)
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/builtins/S3Downloader.java:177]: list the necessary permission and that it is needed in the IAM role.

File: src/main/java/com/aws/greengrass/componentmanager/exceptions/PackageDownloadException.java [hunk 8]
```diff
@@ -8,6 +8,8 @@
 import com.aws.greengrass.deployment.errorcode.DeploymentErrorCode;
 import com.aws.greengrass.deployment.exceptions.DeploymentException;
 
+import java.util.List;
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/exceptions/PackageDownloadException.java:11]: unused

File: src/main/java/com/aws/greengrass/componentmanager/builtins/S3Downloader.java (hunk 6 above)
Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/componentmanager/builtins/S3Downloader.java:130]: If it fails it will throw either an InterruptedException or a proper PackageDownloadException. Moving out of try block so that we don't wrap it with another `PackageDownloadException(getErrorString("Failed to head artifact object from S3"), e)`

File: src/main/java/com/aws/greengrass/componentmanager/builtins/GreengrassRepositoryDownloader.java (hunk 5 above)
Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/componentmanager/builtins/GreengrassRepositoryDownloader.java:224]: will update

File: src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java [hunk 9]
```diff
@@ -167,6 +177,13 @@ private GetDeploymentConfigurationResponse getDeploymentConfiguration(String dep
 
             deploymentConfiguration = greengrassServiceClientFactory.fetchGreengrassV2DataClient()
                     .getDeploymentConfiguration(getDeploymentConfigurationRequest);
+        } catch (AccessDeniedException e) {
+            throw new DeploymentTaskFailureException("getDeploymentConfiguration API returned 403 access denied",
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java:181]: list needed permission, and where it goes (iot policy)

File: src/main/java/com/aws/greengrass/componentmanager/builtins/ArtifactDownloader.java (hunk 4 above)
Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/componentmanager/builtins/ArtifactDownloader.java:138]: will update

File: src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java [hunk 10]
```diff
@@ -194,7 +211,8 @@ private void validateHttpExecuteResponse(HttpExecuteResponse executeResponse)
         //but adding a check as deployment document is read into process memory.
         if (deploymentDocumentSizeOptional.isPresent()
                 && Long.parseLong(deploymentDocumentSizeOptional.get()) > MAX_DEPLOYMENT_DOCUMENT_SIZE_BYTES) {
-            throw new DeploymentTaskFailureException("Exceeded Deployment document size limit, doc ");
+            throw new DeploymentTaskFailureException("Exceeded Deployment document size limit",
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java:214]: tell us what the limit is and how big you're attempting to use

File: src/main/java/com/aws/greengrass/deployment/activator/DeploymentActivator.java [hunk 11]
```diff
@@ -31,10 +34,12 @@ public abstract class DeploymentActivator {
     protected final Kernel kernel;
     protected final DeploymentDirectoryManager deploymentDirectoryManager;
     protected static final Logger logger = LogManager.getLogger(DeploymentActivator.class);
+    protected final ComponentStore componentStore;
 
     protected DeploymentActivator(Kernel kernel) {
         this.kernel = kernel;
         this.deploymentDirectoryManager = kernel.getContext().get(DeploymentDirectoryManager.class);
+        this.componentStore = kernel.getContext().get(ComponentStore.class);
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/activator/DeploymentActivator.java:42]: unused?

File: src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCodeUtils.java [hunk 12]
```diff
@@ -180,38 +179,88 @@ private static void collectErrorCodesFromS3Exception(Set<DeploymentErrorCode> er
...
-        if (matcher.matches()) {
-            String accountId = matcher.group(5);
-            return "aws".equals(accountId);
+        // load component arn from recipe metadata json on disk
+        String arnString;
+        try {
+            arnString = componentStore.getRecipeMetadata(ComponentIdentifier.fromServiceTopics(serviceTopics))
+                    .getComponentVersionArn();
+        } catch (PackageLoadingException e) {
+            logger.atWarn().log("Failed to load component metadata file from disk while classifying component error.");
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCodeUtils.java:230]: no period at the end of logs.

File: src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCodeUtils.java [hunk 13]
```diff
@@ -180,38 +179,88 @@ private static void collectErrorCodesFromS3Exception(Set<DeploymentErrorCode> er
...
         }
 
-        final Matcher matcher = COMPONENT_ARN_PATTERN.matcher(arn);
-        if (matcher.matches()) {
-            String accountId = matcher.group(5);
-            return "aws".equals(accountId);
+        // load component arn from recipe metadata json on disk
+        String arnString;
+        try {
+            arnString = componentStore.getRecipeMetadata(ComponentIdentifier.fromServiceTopics(serviceTopics))
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCodeUtils.java:227]: if the component is locally installed and there is no arn, what happens?

File: src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCodeUtils.java [hunk 14]
```diff
@@ -180,38 +179,88 @@ private static void collectErrorCodesFromS3Exception(Set<DeploymentErrorCode> er
...
+            return DeploymentErrorType.COMPONENT_ERROR;
         }
-        throw new PackageLoadingException("Component arn loaded is not valid");
-    }
 
+        // parse the arn to check if account id is AWS
+        try {
+            Arn arn = Arn.fromString(arnString);
+
+            // sanity check service == greengrass
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCodeUtils.java:238]: why bother with all this? Why not just check for the accountId? This seems unnecessary.

Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCodeUtils.java:238]: True I also wasn't sure if they were necessary. Will remove them.

File: src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCodeUtils.java (hunk 12 above)
Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCodeUtils.java:230]: Will clean up

File: src/main/java/com/aws/greengrass/deployment/activator/DeploymentActivator.java (hunk 11 above)
Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/deployment/activator/DeploymentActivator.java]: It's only used in `DefaultActivator`. Updated

File: src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCodeUtils.java (hunk 13 above)
Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCodeUtils.java:246]: we discussed offline; if a component is locally installed, then we assume it is a user component error

**Review by MikeDombo — DISMISSED** [review]

File: src/main/java/com/aws/greengrass/componentmanager/builtins/ArtifactDownloader.java [hunk 15]
```diff
@@ -131,12 +135,14 @@ public File download() throws PackageDownloadException, IOException, Interrupted
                     offset.set(0);
                     messageDigest.reset();
                     throw new ArtifactChecksumMismatchException(
-                            "Integrity check for downloaded artifact failed. " + "Probably due to file corruption.");
+                            "Integrity check for downloaded artifact failed. Probably due to the file changing"
+                                    + " after creating the component version",
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/componentmanager/builtins/ArtifactDownloader.java:139]: can you change this message? "Probably due to the file changing after creating the component version" is grammatically incorrect.

something like "Failed integrity check for the downloaded artifact. Artifact contents may have changed after component version was created"

File: src/main/java/com/aws/greengrass/componentmanager/builtins/ArtifactDownloader.java [hunk 16]
```diff
@@ -164,7 +170,8 @@ protected long download(InputStream inputStream, MessageDigest messageDigest) th
                 try {
                     artifactFile.write(buffer, 0, readBytes);
                 } catch (IOException e) {
-                    throw new PackageDownloadException(getErrorString("Error writing artifact."), e);
+                    throw new PackageDownloadException(getErrorString("Error writing artifact"), e).withErrorContext(
+                            e.getClass().getSimpleName(), DeploymentErrorCode.IO_WRITE_ERROR);
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/componentmanager/builtins/ArtifactDownloader.java:174]: it would have been useful to add a method `withErrorContext(Throwable t, ErrorCode c)` so you wouldn't need to write `e.getClass().getSimpleName()` everywhere

File: src/main/java/com/aws/greengrass/componentmanager/builtins/GreengrassRepositoryDownloader.java [hunk 17]
```diff
@@ -206,8 +218,21 @@ private String getArtifactDownloadURL(ComponentIdentifier componentIdentifier, S
             }, "get-artifact-size", logger);
         } catch (InterruptedException e) {
             throw e;
+        } catch (GreengrassV2DataException e) {
+            if (e.statusCode() == HttpStatusCode.FORBIDDEN) {
+                throw new PackageDownloadException(getErrorString("Failed to get the pre-signed url of a public or a "
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/componentmanager/builtins/GreengrassRepositoryDownloader.java:223]: do you need to specify Lambda here? Is there a way to know whether it is lambda or not before hand? I expect most customers wouldn't be using lambda so it seems odd to include in the error messages

This also seems very verbose. How about:
`Access denied when calling GetComponentVersionArtifact. Ensure certificate policy grants greengrass:GetComponentVersionArtifact`

File: src/main/java/com/aws/greengrass/componentmanager/builtins/S3Downloader.java [hunk 18]
```diff
@@ -94,6 +96,23 @@ protected long download(long rangeStart, long rangeEnd, MessageDigest messageDig
...
+            if (e.statusCode() == HttpStatusCode.FORBIDDEN) {
+                throw new PackageDownloadException(
+                        getErrorString("Failed to download object from S3. GetObject returns 403 Access Denied"
+                                + "Please check that core device's token exchange role grants s3:GetObject"),
+                        e).withErrorContext(e.getClass().getSimpleName(),
+                        DeploymentErrorCode.S3_GET_BUCKET_ACCESS_DENIED);
+            }
+            if (e.statusCode() == HttpStatusCode.NOT_FOUND) {
+                throw new PackageDownloadException(getErrorString(
+                        "Failed to download object from S3. GetObject returns 404 Resource Not Found"
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/componentmanager/builtins/S3Downloader.java:109]: Similar comments about verbose error messages. I think it would be easier to just say:
"GetObject returned _HTTP ERROR CODE + MSG_. Ensure the IAM role associated with the core device has a policy granting s3:GetObject"

File: src/main/java/com/aws/greengrass/componentmanager/converter/RecipeLoader.java [hunk 19]
```diff
@@ -66,9 +67,9 @@ public static com.amazon.aws.iot.greengrass.component.common.ComponentRecipe par
                     .readValue(recipe, com.amazon.aws.iot.greengrass.component.common.ComponentRecipe.class);
         } catch (JsonProcessingException e) {
             // TODO: [P41216539]: move this to common model
-            throw new PackageLoadingException(
-                    String.format("Failed to parse recipe file content to contract model. Recipe file content: '%s'.",
-                            recipe), e);
+            throw new PackageLoadingException(String.format(
+                    "Failed to parse recipe file content to contract model. Recipe file content: '%s'.", recipe),
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/componentmanager/converter/RecipeLoader.java:71]: should the full recipe contents be part of the exception message? that seems like it should just be logged

File: src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java [hunk 20]
```diff
@@ -137,7 +144,8 @@ private String downloadFromUrl(String deploymentId, String preSignedUrl)
                 executeResponse = client.prepareRequest(executeRequest).call();
             } catch (IOException e) {
                 throw new RetryableDeploymentDocumentDownloadException(
-                        "I/O error when making HTTP request with presigned url.", e);
+                        "I/O error when making HTTP request with presigned url.", e)
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java:147]: nit: remove "." at end of message

File: src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java [hunk 21]
```diff
@@ -147,13 +155,15 @@ private String downloadFromUrl(String deploymentId, String preSignedUrl)
                 return IoUtils.toUtf8String(in);
             } catch (IOException e) {
                 throw new RetryableDeploymentDocumentDownloadException(
-                        "I/O error when reading from HTTP response payload stream.", e);
+                        "I/O error when reading from HTTP response payload stream.", e)
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java:158]: nit: remove "." at end of message

File: src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java [hunk 22]
```diff
@@ -167,6 +177,14 @@ private GetDeploymentConfigurationResponse getDeploymentConfiguration(String dep
 
             deploymentConfiguration = greengrassServiceClientFactory.fetchGreengrassV2DataClient()
                     .getDeploymentConfiguration(getDeploymentConfigurationRequest);
+        } catch (AccessDeniedException e) {
+            throw new DeploymentTaskFailureException("getDeploymentConfiguration API returned 403 access denied. "
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java:181]: nit: GetDeploymentConfiguration to match the IoT policy grant statement. Also similar message as before about format.

File: src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java [hunk 23]
```diff
@@ -167,6 +177,14 @@ private GetDeploymentConfigurationResponse getDeploymentConfiguration(String dep
...
+                    + "Please make sure core device's IoT policy grants greengrass:GetDeploymentConfiguration",
+                    e).withErrorContext(e.getClass().getSimpleName(),
+                    DeploymentErrorCode.GET_DEPLOYMENT_CONFIGURATION_ACCESS_DENIED);
+        } catch (GreengrassV2DataException e) {
+            // TODO: better retry handling
+            throw new DeploymentTaskFailureException("Error while calling getDeploymentConfiguration", e);
         } catch (AwsServiceException e) {
             throw new RetryableDeploymentDocumentDownloadException(
                     "Greengrass Cloud Service returned an error when getting full deployment configuration.", e);
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java:190]: nit: remove "."

File: src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java [hunk 24]
```diff
@@ -209,14 +230,19 @@ private DeploymentDocument deserializeDeploymentDoc(String configurationInString
             Configuration configuration =  SerializerFactory.getFailSafeJsonObjectMapper()
                     .readValue(configurationInString, Configuration.class);
             return DeploymentDocumentConverter.convertFromDeploymentConfiguration(configuration);
-        } catch (Exception e) {
-            throw new DeploymentTaskFailureException("Failed to deserialize deployment document.", e);
+        } catch (IOException e) {
+            throw new DeploymentTaskFailureException("Failed to deserialize deployment document.", e).withErrorContext(
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java:234]: nit: remove "."

File: src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java [hunk 25]
```diff
@@ -209,14 +230,19 @@ private DeploymentDocument deserializeDeploymentDoc(String configurationInString
...
+                    e.getClass().getSimpleName(), DeploymentErrorCode.DEPLOYMENT_DOCUMENT_PARSE_ERROR);
+        } catch (InvalidRequestException e) {
+            throw new DeploymentTaskFailureException("Invalid component metadata from deployment document.",
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java:237]: nit: remove "."

File: src/main/java/com/aws/greengrass/deployment/DeploymentService.java [hunk 26]
```diff
@@ -745,7 +756,8 @@ private DefaultDeploymentTask createDefaultNewDeployment(Deployment deployment)
     private DeploymentDocument parseAndValidateJobDocument(Deployment deployment) throws InvalidRequestException {
         String jobDocumentString = deployment.getDeploymentDocument();
         if (Utils.isEmpty(jobDocumentString)) {
-            throw new InvalidRequestException("Job document cannot be empty");
+            throw new InvalidRequestException("deployment document cannot be empty",
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/deployment/DeploymentService.java:759]: nit: Deployment document - although some error messages say "job document" - when do you use deployment vs job?

File: src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCodeUtils.java (hunk 13 above)
Comment (rbattle) [reply src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCodeUtils.java:246]: is it worth thinking about including the arn in the config store as an attribute and loading from disk only if it doesn't exist?

File: src/main/java/com/aws/greengrass/componentmanager/builtins/ArtifactDownloader.java (hunk 16 above)
Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/componentmanager/builtins/ArtifactDownloader.java:174]: will refactor it

File: src/main/java/com/aws/greengrass/componentmanager/builtins/ArtifactDownloader.java (hunk 15 above)
Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/componentmanager/builtins/ArtifactDownloader.java:139]: will update

File: src/main/java/com/aws/greengrass/componentmanager/builtins/GreengrassRepositoryDownloader.java (hunk 17 above)
Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/componentmanager/builtins/GreengrassRepositoryDownloader.java:223]: will update

File: src/main/java/com/aws/greengrass/componentmanager/builtins/S3Downloader.java (hunk 18 above)
Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/componentmanager/builtins/S3Downloader.java:109]: will update

File: src/main/java/com/aws/greengrass/componentmanager/converter/RecipeLoader.java (hunk 19 above)
Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/componentmanager/converter/RecipeLoader.java:71]: will update

File: src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java (hunk 22 above)
Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java:181]: will update

File: src/main/java/com/aws/greengrass/deployment/DeploymentService.java (hunk 26 above)
Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/deployment/DeploymentService.java:759]: changed to deployment document

File: src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCodeUtils.java (hunk 13 above)
Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCodeUtils.java:248]: We discussed offline; ideally we want to put it under component’s service topic, but the arn is only received during `resolveComponentVersion` , and the greengrass service topic does not exist until we merge config and start services.

Added a TODO to think whether we need to persist it in deployment service or some other place in config.

File: src/main/java/com/aws/greengrass/componentmanager/ComponentManager.java [hunk 27]
```diff
@@ -199,25 +207,22 @@ private ComponentIdentifier negotiateVersionWithCloud(String componentName,
...
-                            versionRequirements);
-                } else {
-                    throw new NoAvailableComponentVersionException(
-                            "No local or cloud component version satisfies the requirements.", componentName,
-                            versionRequirements, e);
-                }
+                throw new NoAvailableComponentVersionException(VERSION_NOT_FOUND_FAILURE_MESSAGE, componentName,
+                        versionRequirements);
+            } catch (AccessDeniedException e) {
+                throw new PackagingException(" Access denied when calling ResolveComponentCandidates. Ensure "
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/componentmanager/ComponentManager.java:221]: nit: Remove the leading space

File: src/main/java/com/aws/greengrass/componentmanager/ComponentManager.java [hunk 28]
```diff
@@ -77,6 +80,11 @@ public class ComponentManager implements InjectionActions {
     private static final long DEFAULT_MIN_DISK_AVAIL_BYTES = 20 * ONE_MB;
     protected static final String COMPONENT_NAME = "componentName";
 
+    public static final String INSTALLED_COMPONENT_NOT_FOUND_FAILURE_MESSAGE =
+            "No active component version satisfies the requirements of non-target groups.";
+    public static final String VERSION_NOT_FOUND_FAILURE_MESSAGE =
+            "No local or cloud component version satisfies the requirements.";
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/componentmanager/ComponentManager.java:86]: nit: remove "."

File: src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java [hunk 29]
```diff
@@ -211,7 +215,8 @@ ComponentRecipe getPackageRecipe(@NonNull ComponentIdentifier pkgId) throws Pack
             // TODO: [P41215929]: Better logging and exception messages in component store
             throw new PackageLoadingException(String.format(
                     "Failed to find usable recipe for current platform: %s, for package: '%s' in the "
-                            + "local package store.", platformResolver.getCurrentPlatform(), pkgId));
+                            + "local package store.", platformResolver.getCurrentPlatform(), pkgId),
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java:218]: nit: remove "."

File: src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java [hunk 30]
```diff
@@ -436,7 +442,8 @@ private static Semver parseVersionFromRecipeFileName(String recipeFilename) thro
             return new Semver(versionStr);
         } catch (SemverException e) {
             throw new PackageLoadingException(
-                    String.format("Component recipe file name: '%s' is corrupted!", recipeFilename), e);
+                    String.format("Component recipe file name: '%s' is corrupted!", recipeFilename), e,
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java:445]: nit: remove "!"

File: src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java [hunk 31]
```diff
@@ -478,7 +486,8 @@ public RecipeMetadata getRecipeMetadata(ComponentIdentifier componentIdentifier)
 
             throw new PackageLoadingException(String.format(
                     "Failed to get recipe metadata because the file doesn't not exit or it is a folder. "
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java:488]: file does not exist or it is a folder

we should be able to distinguish file doesn't exist from a folder though...

also remove the trailing "." - you can see it in a bunch of exceptions in here. We should be consistent and not have trailing "."

File: src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java [hunk 32]
```diff
@@ -496,7 +505,8 @@ public RecipeMetadata getRecipeMetadata(ComponentIdentifier componentIdentifier)
 
             throw new PackageLoadingException(String.format(
                     "Failed to get recipe metadata because the recipe metadata file should be a json but is corrupted."
-                            + " RecipeMetadataFilePath: '%s'.", metadataFile.getAbsolutePath()), e);
+                            + " RecipeMetadataFilePath: '%s'.", metadataFile.getAbsolutePath()), e)
+                    .withErrorContext(e, DeploymentErrorCode.RECIPE_METADATA_PARSE_ERROR);
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java:509]: If it is just invalid JSON then just say something like: "Recipe metadata is not valid JSON: %s" ?

File: src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java [hunk 33]
```diff
@@ -506,7 +516,8 @@ public RecipeMetadata getRecipeMetadata(ComponentIdentifier componentIdentifier)
 
             throw new PackageLoadingException(String.format(
                     "Failed to get recipe metadata because the recipe metadata file json has wrong structure."
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java:518]: do we need to distinguish _every_ type of invalid JSON?

File: src/main/java/com/aws/greengrass/componentmanager/builtins/S3Downloader.java [hunk 34]
```diff
@@ -94,6 +96,19 @@ protected long download(long rangeStart, long rangeEnd, MessageDigest messageDig
...
+        } catch (S3Exception e) {
+            if (e.statusCode() == HttpStatusCode.FORBIDDEN) {
+                throw new PackageDownloadException(getErrorString("S3 GetObject returns 403 Access Denied. "
+                        + "Ensure the IAM role associated with the core device has a policy granting s3:GetObject"),
+                        e).withErrorContext(e, DeploymentErrorCode.S3_GET_BUCKET_ACCESS_DENIED);
+            }
+            if (e.statusCode() == HttpStatusCode.NOT_FOUND) {
+                throw new PackageDownloadException(getErrorString("S3 GetObject returns 404 Resource Not Found."
+                        + "Ensure the IAM role associated with the core device has a policy granting s3:GetObject "
+                        + "and the artifact object url is correct"),
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/componentmanager/builtins/S3Downloader.java:108]: nit: Customers provide a URI

**Review by rbattle — COMMENTED** [review]
more small nits

File: src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java (hunk 33 above)
Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java:518]: cleaned up a bit

**Review by rbattle — APPROVED** [review]

**Review by MikeDombo — APPROVED** [review]
